[pt] Optimize index_select operator

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28865 [pt] Optimize index_select operator**

We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 1.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# python index_select_bench.py
elapsed_per_iter: 0.001158905029296875
GB/s: 8.13
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

jianyuh committed:
[pt] Optimize index_select operator

We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 2.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```

(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.000705552101135254
GB/s: 13.35
```

If we don't replace `simd` with `omp simd` in `aten/src/TH/generic/THTensorApply.hpp`.
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0008904933929443359
GB/s: 10.58
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt] Optimize index_select operator"


We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 2.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```

(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.000705552101135254
GB/s: 13.35
```

If we don't replace `simd` with `omp simd` in `aten/src/TH/generic/THTensorApply.hpp`.
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0008904933929443359
GB/s: 10.58
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

[ghstack-poisoned]

dskhudia commented on aten/src/TH/generic/THTensorApply.hpp:
```diff
@@ -74,7 +74,7 @@ if (std::isnan(val)) break;
 #define PRAGMA(P)         _Pragma(#P)
 #define PRAGMA_IVDEP      PRAGMA(ivdep)
-#define PRAGMA_SIMD       PRAGMA(simd)
+#define PRAGMA_SIMD       PRAGMA(omp simd)
```
This becomes parallelization backend (omp) specific?

jspark1105 commented on aten/src/TH/generic/THTensorEvenMoreMath.cpp:
```diff
@@ -433,10 +433,13 @@ void THTensor_(indexSelect)(THTensor *tensor, THTensor *src, int dim, THLongTens
+            auto cur_row = index_data[i];
+            scalar_t* cur_dst = tensor_data + i * rowsize;
+            scalar_t* cur_src = src_data + cur_row * rowsize;
+            PRAGMA_SIMD
```
I'm surprised that this code is faster than memcpy. Is this because rowsize was small when we benchmark?

jianyuh committed:
Update on "[pt] Optimize index_select operator"


We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 2.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# python index_select_bench.py
elapsed_per_iter: 0.001158905029296875
GB/s: 8.13
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt] Optimize index_select operator"


We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 2.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# python index_select_bench.py
elapsed_per_iter: 0.001158905029296875
GB/s: 8.13
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt] Optimize index_select operator"


We would like to optimize the `index_select` operator in PyTorch for Vocab Reduction models for NMT. With the current PR, we get 1.5x speedup for this operator.

Here is our benchmark (The matrix shapes are from T56133692).

```
import time
import torch

torch.set_num_threads(1)

embedding_num_rows = 27000
embedding_size = 512
indice_size = 2300

embedding_table = torch.rand(embedding_num_rows, embedding_size)
indices = torch.LongTensor(indice_size).random_(0, embedding_num_rows)

WARM_UP_NITER = 5
NITER = 20

def flush_cache():
    my_tensor = torch.rand(int(1e8))
    my_tensor *= 2

for i in range(WARM_UP_NITER):
    flush_cache()
    y = torch.index_select(embedding_table, 0, indices)

elapsed_total_time = 0.0
s = time.time()
for i in range(NITER):
    flush_cache()
    s0 = time.time()
    y = torch.index_select(embedding_table, 0, indices)
    elapsed_per_iter0 = time.time() - s0
    elapsed_total_time += elapsed_per_iter0
elapsed_per_iter = elapsed_total_time / NITER

print("elapsed_per_iter:", elapsed_per_iter)
print("GB/s: {:0.2f}".format(indices.numel()*embedding_size*4*2/elapsed_per_iter/1e9))
```

On single core of Skylake T6:
Before this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]#  numactl --physcpubind=1 --membind=1 python index_select_bench.py
elapsed_per_iter: 0.0017513513565063476
GB/s: 5.38
```

After this Diff:
```
(base) [root@rtptest10054.frc2 ~/jhuang_test/index_select]# python index_select_bench.py
elapsed_per_iter: 0.001158905029296875
GB/s: 8.13
```

More optimizations:
- Add the software prefetching instructions.
- Fuse the index select operator with the following matrix multiplication. Basically, we will provide the general stride matrix multiplication with an additional parameter in `cblas_sgemm`, indices_B, which indicates which rows of B we should choose from. This will take some efforts, but depending on the need from the LATTE team, we can add such general stride matrix multiplication inside FBGEMM.

Differential Revision: [D18220891](https://our.internmc.facebook.com/intern/diff/D18220891/)

[ghstack-poisoned]
